src/header.rs

```diff
@@ -42,6 +42,21 @@ pub struct Header {
     /// Defined in [RFC7515#4.1.7](https://tools.ietf.org/html/rfc7515#section-4.1.7).
     #[serde(skip_serializing_if = "Option::is_none")]
     pub x5t: Option<String>,
+    /// X.509 certificate SHA-256 thumbprint
+    ///
+    /// Defined in [RFC7515#4.1.7](https://tools.ietf.org/html/rfc7515#section-4.1.8).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(rename = "x5t#S256")]
+    pub x5t_s256: Option<String>,
+    /// Critical
+    ///
+    /// Defined in [RFC7515#4.1.11](https://tools.ietf.org/html/rfc7515#section-4.1.11).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub crit: Option<Vec<String>>,
+    /// Additional Public or Private header parameters
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(flatten)]
+    pub params: Option<std::collections::HashMap<String, serde_json::Value>>,
```

Comment:
I'm a bit torn about exposing `Value` here... Can we not assume that it's going to be 99% Strings or are there known usecases where it's not a string? Outside of some homegrown token implementation

It looks like we're missing `x5c` too? I need to go through the spec after merging and ensure we have everything

Reply:
Yes, I forgot `x5c`. [RFC 7515 § 4.1.6](https://tools.ietf.org/html/rfc7515#section-4.1.6) specifies `x5c` as an array of strings (in base64-encoding).

[§ 4.1.11](https://tools.ietf.org/html/rfc7515#section-4.1.11) gives a hypothetical example of a header value that is a number.

[§ 4.1.3](https://tools.ietf.org/html/rfc7515#section-4.1.3) specifies `jwk` is a JWK, which is an object, specified in [RFC 7517](https://tools.ietf.org/html/rfc7517) with its own [registry](https://www.iana.org/assignments/jose/jose.xhtml#web-key-parameters) of parameters - some of which appear to be the same as JOSE header parameters, and several various key parameters.

I am using the `b64` header option (as in #149), which is a boolean.

`crit` is an array of strings.

In the [JWS/JWE header parameters registry](https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-header-parameters), the remaining parameters for JWS not already mentioned are:
- `url` and `nonce`, from [RFC 8555](https://tools.ietf.org/html/rfc8555) (ACME): String
- `ppt`: has its own [registry](https://www.iana.org/assignments/passport/passport.xhtml#passport-extensions) of possible values, which are all strings

I can understand if you don't want to expose Value for arbitrary parameters, as it looks like maybe not the fullest use of the type system. On the other hand, [RFC 7515 § 4.2](https://tools.ietf.org/html/rfc7515#section-4.2) and [§ 4.3](https://tools.ietf.org/html/rfc7515#section-4.3) say that additional header values may be used that are not in the registry (with caution / sparingly), and doesn't place any type restrictions on them. So it would be adhering to the RFC to allow additional arbitrary headers.  With `Value`, a library user could use `serde_json::from_value`/`to_value` to do additional (de)serialization.

If the additonal headers map is not included, would you be open to instead including specific support for additional standardized header? Specifically, I would add `crit` as in this PR, and `b64` (which is easier now that we can sign bytes).

Reply:
I'm thinking adding additional headers should be via a method on `Header` rather than making the field pub. This way we don't need to expose Value, we can just take something that `impl dyn Serialize`. We also need to add another method that returns the HashMap, which would expose Value here, no way around it. But at least we got the serialization taken care of by the library.